File: Day115_2025/Q115_Leetcode.py
from typing import List


class Solution:
    def countInterestingSubarrays(self, nums: List[int], modulo: int, k: int) -> int:
        from collections import defaultdict
        count = defaultdict(int)
        count[0] = 1
        res = 0
        prefix = 0

        for num in nums:
            if num % modulo == k:
                prefix += 1
            key = (prefix - k + modulo) % modulo
            res += count[key]
            count[prefix % modulo] += 1

        return res

File: Day115_2025/Q115_Leetcode.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public long countInterestingSubarrays(List<Integer> nums, int modulo, int k) {
        int n = nums.size();
        Map<Integer, Integer> countMap = new HashMap<>();
        long res = 0;
        int prefix = 0;
        countMap.put(0, 1); // base case

        for (int i = 0; i < n; i++) {
            if (nums.get(i) % modulo == k) {
                prefix++;
            }
            int key = (prefix - k + modulo) % modulo;
            res += countMap.getOrDefault(key, 0);
            int modVal = prefix % modulo;
            countMap.put(modVal, countMap.getOrDefault(modVal, 0) + 1);
        }

        return res;
    }
}
